File: kvstore/store/store.py
"""In-memory key-value store with optional per-key TTL."""

from __future__ import annotations

import threading
import time
from typing import Optional


class KVStore:
    def __init__(self) -> None:
        self._mu = threading.Lock()
        # value → (value, expires_at_monotonic | None)
        self._data: dict[str, tuple[str, float | None]] = {}

    # ── core operations ──────────────────────────

    def get(self, key: str) -> Optional[str]:
        with self._mu:
            entry = self._data.get(key)
            if entry is None:
                return None
            value, expires_at = entry
            if expires_at is not None and time.monotonic() >= expires_at:
                del self._data[key]
                return None
            return value

    def set(self, key: str, value: str, ttl_seconds: Optional[float] = None) -> None:
        expires_at = None
        if ttl_seconds is not None and ttl_seconds > 0:
            expires_at = time.monotonic() + ttl_seconds
        with self._mu:
            self._data[key] = (value, expires_at)

    def delete(self, key: str) -> bool:
        with self._mu:
            if key in self._data:
                del self._data[key]
                return True
            return False

    def keys(self) -> list[str]:
        now = time.monotonic()
        with self._mu:
            return [
                k for k, (_, exp) in self._data.items()
                if exp is None or exp > now
            ]

    def size(self) -> int:
        return len(self.keys())

    # ── Raft command application ─────────────────

    def apply(self, command: dict) -> Optional[str]:
        """
        Apply a committed Raft command to the store.
        Supported ops: "set", "delete".
        Returns the stored value for "set", None otherwise.
        """
        op = command.get("op")
        key = command.get("key", "")
        if op == "set":
            value = command.get("value", "")
            ttl = command.get("ttl")
            self.set(key, value, ttl_seconds=ttl)
            return value
        if op == "delete":
            self.delete(key)
            return None
        return None

    # ── TTL sweep ────────────────────────────────

    def purge_expired(self) -> int:
        """Remove all expired keys. Returns the number of keys removed."""
        now = time.monotonic()
        with self._mu:
            expired = [
                k for k, (_, exp) in self._data.items()
                if exp is not None and now >= exp
            ]
            for k in expired:
                del self._data[k]
            return len(expired)

File: store/store.go
package store

import (
	"sync"
	"time"
)

// Command is the committed operation format shared by the store, WAL, and Raft.
type Command map[string]any

type entry struct {
	value     string
	expiresAt time.Time
}

// KVStore is an in-memory key-value store with optional per-key TTL.
type KVStore struct {
	mu   sync.Mutex
	data map[string]entry
}

func New() *KVStore {
	return &KVStore{data: make(map[string]entry)}
}

func (s *KVStore) Get(key string) (string, bool) {
	s.mu.Lock()
	defer s.mu.Unlock()

	e, ok := s.data[key]
	if !ok {
		return "", false
	}
	if !e.expiresAt.IsZero() && !time.Now().Before(e.expiresAt) {
		delete(s.data, key)
		return "", false
	}
	return e.value, true
}

func (s *KVStore) Set(key, value string, ttl time.Duration) {
	var expiresAt time.Time
	if ttl > 0 {
		expiresAt = time.Now().Add(ttl)
	}

	s.mu.Lock()
	s.data[key] = entry{value: value, expiresAt: expiresAt}
	s.mu.Unlock()
}

func (s *KVStore) Delete(key string) bool {
	s.mu.Lock()
	defer s.mu.Unlock()

	if _, ok := s.data[key]; !ok {
		return false
	}
	delete(s.data, key)
	return true
}

func (s *KVStore) Keys() []string {
	now := time.Now()

	s.mu.Lock()
	defer s.mu.Unlock()

	keys := make([]string, 0, len(s.data))
	for k, e := range s.data {
		if e.expiresAt.IsZero() || now.Before(e.expiresAt) {
			keys = append(keys, k)
		}
	}
	return keys
}

func (s *KVStore) Size() int {
	return len(s.Keys())
}

// Apply applies a committed Raft command. Supported ops are set and delete.
func (s *KVStore) Apply(command Command) (string, bool) {
	op, _ := command["op"].(string)
	key, _ := command["key"].(string)

	switch op {
	case "set":
		value, _ := command["value"].(string)
		s.Set(key, value, ttlFromCommand(command["ttl"]))
		return value, true
	case "delete":
		s.Delete(key)
	}
	return "", false
}

func (s *KVStore) PurgeExpired() int {
	now := time.Now()

	s.mu.Lock()
	defer s.mu.Unlock()

	removed := 0
	for k, e := range s.data {
		if !e.expiresAt.IsZero() && !now.Before(e.expiresAt) {
			delete(s.data, k)
			removed++
		}
	}
	return removed
}

func ttlFromCommand(v any) time.Duration {
	switch t := v.(type) {
	case nil:
		return 0
	case time.Duration:
		return t
	case float64:
		return time.Duration(t * float64(time.Second))
	case float32:
		return time.Duration(float64(t) * float64(time.Second))
	case int:
		return time.Duration(t) * time.Second
	case int64:
		return time.Duration(t) * time.Second
	case int32:
		return time.Duration(t) * time.Second
	default:
		return 0
	}
}
